parent_visitor.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2017 Ernest Micklei
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package proto
+
+// ParentAwareVisitor keeps track of all the parents of each AST node
+// while visiting all the elements and options.
+type ParentAwareVisitor struct {
+	Parents  []Visitee
```

Comment:
It might be nice to have `LastParent Visitee` which is `nil` if there is no parent, otherwise has the last element of the `Parents` list, checking `if LastParent != nil` is nicer than `len(Parents) > 0` and then getting the last parent with `Parents[len(Parents)-1]`

Reply:
Ah, actually disregard potentially, if you have https://github.com/emicklei/proto/pull/38 it's fine.

Reply:
thinking about dropping #38 in favour of this.